File: 2024/1.py
"""
Day 1 of Advent of Code 2024
"""

with open("1.example", "r") as f:
    example_data = f.read().splitlines()

with open("1.data", "r") as f:
    data = f.read().splitlines()

def parse_lists(lines: list[str]) -> tuple[list[int], list[int]]:
    left = []
    right = []
    for line in lines:
        li, *_, ri = line.split(' ')
        left.append(int(li))
        right.append(int(ri))
    return left, right

def solve_part_one(lines: list[str]) -> int:
    total = 0
    left, right = parse_lists(lines)
    left = sorted(left)
    right = sorted(right)

    for i in range(len(left)):
        distance = abs(left[i] - right[i])
        total += distance
    return total

assert solve_part_one(example_data) == 11

print(solve_part_one(data))

def solve_part_two(lines: list[str]) -> int:
    score = 0
    left, right = parse_lists(lines)
    for item in left:
        score += item * right.count(item)
    return score

assert solve_part_two(example_data) == 31

print(solve_part_two(data))

File: 2024/1.rs
use std::fs;

fn main() {
    let example = fs::read_to_string("../1.example").expect("File error");
    let result = solve_part_one(example.clone());
    println!("Part 1 - Example: {}", result); // 11
    let data = fs::read_to_string("../1.data").expect("File error");
    let result = solve_part_one(data.clone());
    println!("Part 1 - Real: {}", result); // 1603498

    let result = solve_part_two(example);
    println!("Part 2 - Example: {}", result); // 31
    let result = solve_part_two(data);
    println!("Part 2 - Real: {}", result); // 25574739
}

fn solve_part_one(contents: String) -> i32 {
    let lines = contents.trim().lines();
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();
    let mut total = 0;
    for line in lines {
        let items: Vec<&str> = line.split(" ").collect();
        left.push(items[0].parse().unwrap());
        right.push(items[3].parse().unwrap());
    }
    left.sort_unstable();
    right.sort_unstable();
    for (i, item) in left.iter().enumerate() {
        let result = (item - right[i]).abs();
        total += result;
    }
    return total;
}

fn solve_part_two(contents: String) -> i32 {
    let lines = contents.trim().lines();
    let mut left: Vec<i32> = Vec::new();
    let mut right: Vec<i32> = Vec::new();
    let mut total = 0;
    for line in lines {
        let items: Vec<&str> = line.split(" ").collect();
        left.push(items[0].parse().unwrap());
        right.push(items[3].parse().unwrap());
    }
    for item in left.iter() {
        let count: i32 = right.iter().filter(|x| x == &item).count().try_into().unwrap();
        total += item * count;
    }
    return total;
}
